docs/writing-rules/experiments/symbolic-propagation.md

```diff
@@ -27,3 +27,24 @@ If we tried to match the pattern `pandas.DataFrame(...).index.set_value(...)` ag
 Symbolic propagation is a generalization of [constant propagation](/writing-rules/data-flow/constant-propagation/) that addresses this limitation. It enables Semgrep to perform matching modulo variable assignments. Thus, Semgrep is then able to match both `test1` and `test2` with the same simple pattern. This feature needs to be enabled explicitly via rule `options:` by setting `symbolic_propagation: true`.
 
 <iframe src="https://semgrep.dev/embed/editor?snippet=JeBP" border="0" frameBorder="0" width="100%" height="432"></iframe>
+
+## Limitations
+
+Currently, symbolic propagation does not cross branching boundaries, such as `if` clauses or loops. This helps avoid difficulties with managing the parse graph for the code. Consider the following Python code, adapted from the example shown above:
+
+```python
+import pandas
+
+def test1():
+    # ruleid: test
+    pandas.DataFrame(x).index.set_value(a, b, c)
+
+def test2():
+    if (x < 5):
+        df = pandas.DataFrame(x)
+    ix = df.index
+    # ruleid: test
+    ix.set_value(a, b, c)
+```
+
+In this case, even if `symbolic_propagation: true` is used, Semgrep does not match `test2`, because the assignment of `df` to `pandas.DataFrame(x)` is within a conditional, and is not propagated to the final two lines.
```

Comment:
I would use this example instead:
```python
 def test2():
     df = pandas.DataFrame(x)
     if (x < 5):
        pass
     ix = df.index
     # ruleid: test
     ix.set_value(a, b, c)
 ```
So even if `df` is defined outside the `if` block, its symbolic value is lost after the `if` block.

Reply:
I see, is that to show that it doesn't have to be within the conditional, there just has to be a conditional present? That is a more general limitation, so it makes sense to me, just want to confirm.

Reply:
Yes, any symbolic value that is known before the `if (x < 5)`, or introduced in its body, will be lost after the `if`. The control-flow graph has "join" points, where two or more branches get "merged"/"joined". At those "join" points, right now Semgrep forgets every symbolic-propagated value. As of now symbolic values are only propagated through sequential control-flow.

Reply:
I'll try to relax this restriction soon, maybe next week, so maybe we need to update the docs soon. 🤷‍♂️

Reply:
If it's that soon maybe it's not worth merging! I wasn't sure how long it might be. If there will still be a restriction of some sort we can go ahead and then patch it later, though.

Reply:
I can't make promises unless that issues becomes high-prio, and the work here is done I think, so I would just merge this.